### PR TITLE
feat(ethereum): add ABI type awareness with proxy->implementation resolution

### PR DESCRIPTION
**Pull request by pepe-anchor**
## Why

ABI enrichment for Ethereum assumed the transaction's `to` address was always the implementation contract. For a proxy that's wrong: the proxy delegates to an implementation, so the calldata selector belongs to the implementation's ABI, not the proxy's. Today a wallet has no way to tell the parser "this address is a proxy, decode against this other implementation ABI", so proxied calls fall back to raw hex.

This adds a way for wallets to declare the ABI type (proxy vs implementation, extensible to multicall etc. later) and, for proxies, link to the implementation ABI used for decoding.

## What

- **Proto** (`Abi` message): new `AbiType` enum (`UNSPECIFIED`/`IMPLEMENTATION`/`PROXY`) and optional `abi_type` + `implementation_address` fields. Additive and backward compatible.
- **Decoding**: the ABI registry carries an `AbiKind` + implementation link per address. When a tx targets a registered proxy, calldata is decoded against the linked implementation's ABI, falling back to the proxy's own ABI, then hex. The "To" field gets a `Proxy` badge and an `Implementation` address field is shown.
- **Serde**: `abi_type` travels as its string name (`"ABI_TYPE_PROXY"`) over JSON (wallet-facing gateway path), via a `serde(with)` field attribute + `abi_type_serde` module in the generated crate. gRPC carries the enum natively.
- **CLI**: new `--abi-proxy-mappings 0xProxy:0xImpl` flag; the existing `Name:/path:0xAddr` format is untouched.

## Security

- Proxy resolution lives strictly **after** the known-token short-circuit, so a caller-supplied "proxy" entry can never redirect decoding for a canonical-token address (USDC etc.). Regression-tested.
- `abi_type` / `implementation_address` are **not** covered by the ABI signature (which hashes `abi.value` only). This is acceptable: the whole caller-ABI decode path is already untrusted and display-only (only the ABI JSON content is signed, never the address binding), and an attacker who can tamper metadata could already swap the bound ABI. The full `ChainMetadata` is still committed to by `metadata_digest` in the enclave output. Documented in the `abi_metadata` module security notes.

## Backwards compatibility

- Proto change is additive (`optional` fields). Existing clients that don't send `abi_type` keep working; the field defaults to absent and is treated as implementation.
- `AbiRegistry::map_address` is preserved (defaults to implementation), so compile-time embedded ABI callers are unchanged.
- **Heads-up**: adding fields to the borsh-derived `Abi` shifts `metadata_digest` for the same logical ABI across parser versions, even when the new fields are unset. Confirm nothing on the Turnkey side pins/compares `metadata_digest` across versions before rollout. Requires a fresh enclave deployment regardless (it's a binary change).

## Rollback

Revert this commit and redeploy the parser binary. The proto change is additive, so no client coordination is needed to roll back; clients that started sending `abi_type` simply have it ignored by the older binary.

## Test evidence

- `visualsign-ethereum`: 218 unit tests pass (+ new `abi_registry`/`abi_metadata` proxy tests). Note: one pre-existing failure, `test_known_token_short_circuit_uses_tx_chain_id_not_metadata`, fails on `main` too (predates this branch; unrelated to this change).
- `lib_test` e2e: 9/9, including proxy-decodes-via-implementation and the canonical-token-cannot-be-overridden security test.
- `parser_gateway`: serde string round-trip (deserialize / serialize / default / reject-unknown) 11/11.
- `parser_cli`: `--abi-proxy-mappings` flag tests pass. `parser_app`: 5/5 (digest determinism holds). `integration` (against built binaries): 17 pass.
- `cargo fmt --check` clean; `cargo clippy --all-targets -- -D warnings` clean; `make generated` produces no diff.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/parser/cli/src/ethereum.rs [hunk 1]
```diff
@@ -45,40 +52,87 @@ impl crate::ChainPlugin for EthereumPlugin {
...
+        let entry = abi_mappings.entry(proxy.to_string()).or_insert_with(|| {
+            eprintln!("  Note: proxy '{proxy}' has no ABI file; synthesizing an empty proxy ABI");
+            Abi {
+                value: "[]".to_string(),
+                signature: None,
+                ..Default::default()
+            }
+        });
+        entry.abi_type = Some(AbiType::Proxy as i32);
+        entry.implementation_address = Some(implementation.to_string());
```
[the remark this answers is not in the document]
Comment (pepe-anchor) [reply src/parser/cli/src/ethereum.rs]: Good catch. Added `normalize_eth_address` which lowercases the hex portion after `0x`, then applied it in both `build_abi_mappings_from_files` (when inserting map keys) and `apply_proxy_mappings` (when looking up the proxy key and storing the implementation address). The `validate_eth_address` call already runs before normalization so the unwrap inside normalize is safe. Added a regression test (`test_proxy_mapping_mixed_case_links_correctly`) that supplies uppercase in `--abi-json-mappings` and lowercase in `--abi-proxy-mappings` for the same address pair and asserts the proxy retains its ABI file content rather than a synthesized empty one.

---
*Addressed by Claude, code change applied. React 👎 if this doesn't land and I'll flag it for follow-up.*

File: src/parser/cli/src/ethereum.rs [hunk 2]
```diff
@@ -232,28 +324,199 @@ mod tests {
...
+            proxy_abi.value.contains("upgradeTo"),
+            "proxy ABI should contain the file content, not the synthesized empty ABI"
+        );
+        assert_eq!(
+            proxy_abi.implementation_address.as_deref(),
+            Some("0xbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb")
+        );
+
+        // The baseline case still works (lowercase in both flags).
+        let _ = (abi_mappings, proxy_links);
```
[the remark this answers is not in the document]
Comment (pepe-anchor) [reply src/parser/cli/src/ethereum.rs]: Removed the dead code (the unused `proxy_path`, `impl_path`, `abi_mappings`, and `proxy_links` locals from lines 433-451) and the `let _ = (abi_mappings, proxy_links)` suppressor. The test now only contains the code that actually exercises the scenario: uppercase addresses in `--abi-json-mappings` paired with lowercase addresses in `--abi-proxy-mappings`.

---
*Addressed by Claude, code change applied. React 👎 if this doesn't land and I'll flag it for follow-up.*

**Review by shahan-khatchadourian-anchorage — COMMENTED** [review]
Overall well-shaped. Proxy resolution is single-hop, casing is normalized in both paths, the known-token short-circuit sits correctly ahead of the caller-ABI path, and the `metadata_digest` shift is acknowledged in the commit body. None of the items below are blockers — flagging for consideration.

### 1. Proxy fallback path is internally inconsistent

`lib.rs:535-541`: when the impl ABI is missing or doesn't decode, the code falls through to the proxy's own ABI. In that fallback the **`Implementation` field is not emitted**, but the **`Proxy` badge is still emitted** — `lib.rs:591-595` checks only `abi_kind`. Consider always emitting `Implementation` whenever the destination is `AbiKind::Proxy`, with badge text like `"Proxy implementation (unresolved)"` when the impl ABI didn't drive decoding.

### 2. Unused `Default` derive on `AbiKind`

`abi_registry.rs:25-33` derives `Default` but every call site writes `AbiKind::Implementation` explicitly. Drop the derive or add a one-line justification.

### 3. Test coverage gaps

- **Proxy fallback decode**: proxy with `implementation_address` pointing at an unregistered impl (`lib.rs:543-548`) is unexercised end-to-end.
- **Single-hop semantics**: A → B where B is also `Proxy`, or A → A. Pin the single-hop behavior with a test.
- **Impl ABI present but selector mismatch**: verify fallback to proxy's own ABI.

### 4. CLI `apply_proxy_mappings` edge case

`cli/src/ethereum.rs:150-159`: if the proxy ABI file load failed earlier, `apply_proxy_mappings` silently synthesizes an empty `[]` ABI. Detect that the proxy address appeared in `--abi-json-mappings` and emit a louder warning when the synthetic replaces a failed load.

### 5. Misnamed fixture

`cli/src/ethereum.rs:278`: `"Uniswap:...:0xdAC17F958D2ee523a2206206994597C13D831ec7"` uses the USDT mainnet address. Rename `Uniswap` → `Token`.

**Review by shahan-khatchadourian-anchorage — DISMISSED** [review]
no blockers from my end. my comments above are not blockers. lgtm

**Review by pepe-anchor — COMMENTED** [review]
All five items addressed in b93a6ab:

**1. Proxy fallback inconsistency** — `unresolved_implementation_field` is now emitted whenever `AbiKind::Proxy` is set and the implementation ABI exists but doesn't decode the selector. Also fixed a follow-on: the unresolved marker was making `input_fields` non-empty, silently gating out the raw-hex fallback — added an explicit `FallbackVisualizer` append in the `else` branch so the signer always sees the calldata bytes.

**2. Unused `Default` on `AbiKind`** — derive and `#[default]` attribute removed; doc comment updated to reference `resolve_abi_kind` explicitly.

**3. Test coverage** — three tests added:
- `test_proxy_with_unregistered_impl_produces_no_implementation_field`
- `test_proxy_impl_abi_selector_mismatch_emits_unresolved_implementation`
- `test_proxy_resolution_is_single_hop` (A→B where B is also `Proxy`)

**4. CLI `apply_proxy_mappings` edge case** — pre-compute the set of addresses attempted in `--abi-json-mappings`; emit `Warning:` instead of `Note:` when the synthesized empty ABI replaces a file that was specified but failed to load.

**5. Fixture rename** — `Uniswap` → `Token`.

**Review by shahan-khatchadourian-anchorage — APPROVED** [review]
Thanks @pepe-anchor lgtm
